### PR TITLE
[#3686] Add new msys distribution with updated Git.

### PR DESCRIPTION
**Pull request by alibotean**
Scope
=====

Update Git for Windows development to latest version 2.11.0

Changes
=======

I've built a new distribution with the latest Git version. A lot of changes happened from the last attempt made by @cristi.

Git for Windows it's using a forked MSYS2 internally which specifically disables packet mangement. In other words there is no easy way to install packages that are not already included. Luck has it that we only need `wget` to be able to run the development environment. After a lot of searched I've found a wget binary that works and included it as well.

Distribution is based on:
* Portable Git for Windows 2.11.0 - https://git-scm.com/download/win
* wget 1.18 - https://eternallybored.org/misc/wget/

We're using the 32 bit versions.

---

Unfortunately even though we patch the `PATH` for the spawned bash process to include `chevah\git\bin` this does not work for the current release because on start bash spawns it's own process and does not copy over the existing PATH variable.

How to try and test the changes
===============================

reviewers: @adiroiban 

I've tested it on my machine and it will download the dist on first call, then just execute the included Git `bash`.

### CODE REVIEW

**Comment by alibotean** [comment]
Fire away 😃 

needs-review

File: msys-console.js [hunk 1]
```diff
@@ -1,332 +1,320 @@
...
+
+To create the zip archive that is stored in the binary server :
+- decompress Portable Git For Windows (https://git-scm.com/download/win)
+- decompress wget binary and place it in Git\bin folder (https://eternallybored.org/misc/wget/)
+- create a zip archive from git folder
+- upload the zip archive to the server
+- modify the archive name in this script
+*/
+
+var BINARIES_ROOT_URL = 'http://binary.chevah.com/production/git4windows'
```
Comment (adiroiban) [inline msys-console.js]: why a new folder?

**Comment by adiroiban** [comment]
diff is hard to follow... it would help if the newfile fix will be merged in master
and there here to only check the real diff :)

**Comment by alibotean** [comment]
> diff is hard to follow.

fixed

File: msys-console.js [hunk 2]
```diff
@@ -15,29 +15,24 @@ The following times when the script is run, it will see that git was
...
 
 To create the zip archive that is stored in the binary server :
-- install git using the installer
-- install msys using the installer
-- install msys-wget package under msys and copy the files to git folder
-(mingw-get install msys-get, the required files are installed under msys
-folder; the msys-get.exe and required dll dependencies like msys-crypto,
-msys-ssl, msys-intl must be copied to git)
+- decompress Portable Git For Windows (https://git-scm.com/download/win)
+- decompress wget binary and place it in Git\bin folder (https://eternallybored.org/misc/wget/)
```
Comment (adiroiban) [inline msys-console.js]: I remember that you talked about git using a forked msys... and other details, but I don't see that info here.

someone might ask why get wget from there and not us msys get

Comment (adiroiban) [inline msys-console.js]: maybe link to http://stackoverflow.com/questions/25019057/how-are-msys-msys2-and-msysgit-related-to-each-other

Comment (adiroiban) [inline msys-console.js]: in the upstream git commit explain the switch to msys2 there is a comment that pacman works in git for windows?

is that not the case?

File: msys-console.js [hunk 3]
```diff
@@ -15,29 +15,24 @@ The following times when the script is run, it will see that git was
...
 - create a zip archive from git folder
 - upload the zip archive to the server
 - modify the archive name in this script
-
 */
 
-var BINARIES_ROOT_URL = 'http://binary.chevah.com/production/msys-git'
+var BINARIES_ROOT_URL = 'http://binary.chevah.com/production/git4windows'
```
Comment (adiroiban) [inline msys-console.js]: maybe we should name it chevah-git... or something like that... it is not git for windows

File: msys-console.js [hunk 4]
```diff
@@ -309,14 +300,12 @@ function run() {
...
-    var mingw_chevah_root = convert_path_to_mingw_format(chevah_root)
-
-    var cmd = '\"' + mingw_root + '\\git\\bin\\bash.exe\" -c \"' +
-            'PATH=$PATH:' + mingw_chevah_root + '/mingw/git/bin' +
-            ' bash -i -l\"'
+    set_process_env_var('PATH', chevah_root + '\\git\\bin' + path)
 
+    var cmd = '\"' + chevah_root + '\\git\\bin\\bash.exe\" -i -l'
+    
+    //log('cmd = ' + cmd)
```
Comment (adiroiban) [inline msys-console.js]: why this commented code?

File: msys-console.js (hunk 2 above)
Comment (alibotean) [inline msys-console.js]: > I remember that you talked about git using a forked msys... and other details, but I don't see that info here.

We can add it but I prefer to put a link to the project, as they might decide to change in the future and make this all irrelevant -- just as it happend to me when I've tried to go the route described here.

File: msys-console.js (hunk 3 above)
Comment (alibotean) [inline msys-console.js]: ok

File: msys-console.js (hunk 4 above)
Comment (alibotean) [inline msys-console.js]: I was used for debugging only, we don't need it. I'd rather leave it there in case we run into some issues on the dev machines. It will be easier to uncomment this line then to add it yourself so we know what path is used.

**Comment by alibotean** [comment]
> looks good... but where is the code / info about the profile stuff?

I haven't understood exactly what you expect from the `profile`, this is the code as it was only patched to work with the updated version.

We can decide how to handle profile and implement it here or in a separate PR.

File: msys-console.js (hunk 2 above)
Comment (alibotean) [inline msys-console.js]: Managed to find it: http://stackoverflow.com/questions/32712133/package-management-in-git-for-windows

Comment (adiroiban) [inline msys-console.js]: ok. but I think that is imporant to note that wget don't work with pacman ... or pacman don't work with gitmsys ... or what is the root of this evel

File: msys-console.js (hunk 3 above)
Comment (adiroiban) [inline msys-console.js]: or rename it msys-console, as this script has the  same name

**Comment by adiroiban** [comment]
so, with the current script, I can get a working git, but there is no wget

so if I do `./paver` on a fresh repo, it will fail to get the files using wget

from the comment script, I don't know what I did wrong... I don't see where wget is and I was expecting that it will also provide me a wget

needs-changes

**Comment by alibotean** [comment]
> o, with the current script, I can get a working git, but there is no wget
> so if I do ./paver on a fresh repo, it will fail to get the files using wget
> from the comment script, I don't know what I did wrong... I don't see where wget is and I was expecting that it will also provide me a wget

That's what I was trying to say with the `PATH` issue. `wget` is installed in `chevah\git\bin` but that folder is not in the path of the `bash` process.

* If I manage to fix the path do we still want an auto generated `profile` file? 
* If I don't manage to do it; would generating a `profile` file that fixes the PATH and adds an entry for loading `.bash_rc` can be considered a viable option?

**Comment by alibotean** [comment]
needs-review

**Comment by adiroiban** [comment]
why not put wget in /usr/local/bin or one of the values which are already part of PATH?

**Comment by adiroiban** [comment]
I don't what is here

we have
* /chevah/git/bin
* /chevah/git/cmd
* /chevah/usr/bin
* /chevah/mingw32/bin

also, why not call git-bash.exe as recomended in the git portable readme ?

also, I think that force_cscirpt should only be called for debugging, as otherwise I am left with an extra window

needs-changes

**Comment by alibotean** [comment]
> also, why not call git-bash.exe as recomended in the git portable readme ?

That is a no-go option. I've tried it first but it does not work with `python` because of the way it handles input/output. To be able to use `git-bash` you need to call `python` via `winpty` -- I could not find an easy and transparent way to do this; hence I went for `bash`.

**Comment by alibotean** [comment]
> I don't what is here
> we have

We have exactly what we used to have up until now; nothing was changed.

**Comment by alibotean** [comment]
> also, I think that force_cscirpt should only be called for debugging, as otherwise I am left with an extra window

I'm fixing it.

**Comment by adiroiban** [comment]
> I could not find an easy and transparent way to do this; hence I went for bash.

I think that we need a commend where we call bash to explain that git-bash is not working and why.
We need to document all the pay you endure while trying to make it work.

**Comment by alibotean** [comment]
We got improved colors, customized PS1 prompt and no dangling windows left. I'm for leaving the path as is and moving the wget into `/usr/bin`. I will fix this in the distribution and update it on the server.

**Comment by alibotean** [comment]
needs-review

File: msys-console.js [hunk 5]
```diff
@@ -324,9 +378,8 @@ var rc = 0
 
 // The script needs to be run with cscript in order to have all the output
 // redirected to the console and not the GUI, so we enforce that here.
-force_cscript()
+// force_cscript()
```
Comment (adiroiban) [inline msys-console.js]: this line needs to removed, or call it only when a var DEBUG = true is present

we don't want dirty code

**Comment by adiroiban** [comment]
I ok with leaving path as it is and put wget in one of the bin folders

Thanks!

changes-approved

**Comment by alibotean** [comment]
Distribution was updated on Deag.
